types/tablet_v2/wlr_tablet_v2.c

```diff
@@ -0,0 +1,310 @@
+#ifndef _POSIX_C_SOURCE
+#define _POSIX_C_SOURCE 200809L
+#endif
+
+#include <assert.h>
+#include <libinput.h>
+#include <stdlib.h>
+#include <string.h>
+#include <wayland-server.h>
+#include <types/wlr_tablet_v2.h>
+#include <wlr/config.h>
+#include <wlr/types/wlr_seat.h>
+#include <wlr/types/wlr_tablet_tool.h>
+#include <wlr/types/wlr_tablet_v2.h>
+#include <wlr/util/log.h>
+#include "tablet-unstable-v2-protocol.h"
+
+#define TABLET_MANAGER_VERSION 1
+
+struct wlr_tablet_manager_client_v2 {
+	struct wl_list link;
+	struct wl_client *client;
+	struct wl_resource *resource;
+	struct wlr_tablet_manager_v2 *manager;
+
+	struct wl_list tablet_seats; // wlr_tablet_seat_client_v2::link
+};
+
+static void handle_wlr_seat_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_tablet_seat_v2 *seat =
+		wl_container_of(listener, seat, seat_destroy);
+
+	wl_list_remove(&seat->link);
+	wl_list_remove(&seat->seat_destroy.link);
+
+	struct wlr_tablet_seat_client_v2 *client;
+	struct wlr_tablet_seat_client_v2 *tmp;
+	wl_list_for_each_safe(client, tmp, &seat->clients, seat_link) {
+		wlr_tablet_seat_client_v2_destroy(client->resource);
+	}
+}
+
+static struct wlr_tablet_seat_v2 *create_tablet_seat(
+		struct wlr_tablet_manager_v2 *manager,
+		struct wlr_seat *wlr_seat) {
+	struct wlr_tablet_seat_v2 *tablet_seat =
+		calloc(1, sizeof(struct wlr_tablet_seat_v2));
+	if (!tablet_seat) {
+		return NULL;
+	}
+
+	tablet_seat->manager = manager;
+	tablet_seat->wlr_seat = wlr_seat;
+
+	wl_list_init(&tablet_seat->clients);
+
+	wl_list_init(&tablet_seat->tablets);
+	wl_list_init(&tablet_seat->tools);
+	wl_list_init(&tablet_seat->pads);
+
+	tablet_seat->seat_destroy.notify = handle_wlr_seat_destroy;
+	wl_signal_add(&wlr_seat->events.destroy, &tablet_seat->seat_destroy);
+
+	wl_list_insert(&manager->seats, &tablet_seat->link);
+	return tablet_seat;
+}
+
+struct wlr_tablet_seat_v2 *get_or_create_tablet_seat(
+		struct wlr_tablet_manager_v2 *manager,
+		struct wlr_seat *wlr_seat) {
+	struct wlr_tablet_seat_v2 *pos;
+	wl_list_for_each(pos, &manager->seats, link) {
+		if (pos->wlr_seat == wlr_seat) {
+			return pos;
+		}
+	}
+
+	return create_tablet_seat(manager, wlr_seat);
+}
+
+static void tablet_seat_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct zwp_tablet_seat_v2_interface seat_impl = {
+	.destroy = tablet_seat_destroy,
+};
+
+struct wlr_tablet_seat_client_v2 *tablet_seat_client_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwp_tablet_seat_v2_interface,
+		&seat_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+void wlr_tablet_seat_client_v2_destroy(struct wl_resource *resource) {
```

Comment:
I think we normally have two different functions for this, where one takes a resource and the other takes the object.

Reply:
Yeah, I'm not a fan of exposing functions which take a generic `wl_resource` as an argument.

Reply:
Yea, this one should be more internal, it's in the right header, but should have the wlr prefix dropped.